### PR TITLE
Add prefix argument in loggers

### LINKED ISSUE #4537: prefix argument in loggers
## 🚀 Feature
A simple `prefix` argument in loggers, similar to what we have in [ModelCheckpoint](https://github.com/PyTorchLightning/pytorch-lightning/blob/51cc7a89ee9f1c0ad35c147ce37a5825ec6d77e5/pytorch_lightning/callbacks/model_checkpoint.py#L145) that can prepend this value in metric names.

### Motivation
One use-case where I need this is while doing Kfold training. For now, I have to do it manually by updating the metric_name in `self.log`, but this can be done by the loggers using this `prefix` parameter. There might be other use-cases too.

### Pitch
if `prefix='fold1'`
and we do self.log('val_loss', ...)
logged metric will be `fold1_val_loss`.

Let me know if this is a good feature to add :)

cc @PyTorchLightning/core-contributors

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes #4537

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In in short, see following bullet-list:  

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by rohitgr7** [comment]
not sure why some tests are getting cancelled

File: pytorch_lightning/loggers/base.py [hunk 1]
```diff
@@ -305,6 +305,12 @@ def name(self) -> str:
     def version(self) -> Union[int, str]:
         """Return the experiment version."""
 
+    def _add_prefix(self, metrics: Dict[str, float], prefix: str, join_char: str):
```
Comment (SkafteNicki) [inline pytorch_lightning/loggers/base.py:308]: Any reason why `prefix` and `join_char` is given as function arguments when they are part of self?

Comment (rohitgr7) [reply pytorch_lightning/loggers/base.py]: no reason. Updated. Thanks.

**Review by tchaton — APPROVED** [review]
Great addition !

**Review by SkafteNicki — APPROVED** [review]
LGTM

**Comment by rohitgr7** [comment]
@awaelchli do you know why some tests here are getting canceled here?? Something wrong with mock tests here?

**Review by s-rog — APPROVED** [review]

**Comment by edenlightning** [comment]
@tchaton make sure tests here are good?

File: tests/loggers/test_all.py [hunk 2]
```diff
@@ -316,3 +316,67 @@ def _test_logger_created_on_rank_zero_only(tmpdir, logger_class):
...
+    TestTubeLogger,
+    WandbLogger,
+])
+def test_logger_with_prefix_all(tmpdir, monkeypatch, logger_class):
+    _patch_comet_atexit(monkeypatch)
+    try:
+        _test_loggers_with_prefix(tmpdir, logger_class)
+    except (ImportError, ModuleNotFoundError):
+        pytest.xfail(f"pickle test requires {logger_class.__class__} dependencies to be installed.")
+
```
Comment (awaelchli) [inline tests/loggers/test_all.py:335]: @rohitgr7 Pretty sure this is not the correct way to handle this test. 
See the other tests, we need to mock the experiment dependency inside the logger, since we can't install it in CI.

**Review by awaelchli — CHANGES_REQUESTED** [review]

File: tests/loggers/test_all.py (hunk 2 above)
Comment (rohitgr7) [reply tests/loggers/test_all.py:335]: will check this.

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 3]
```diff
@@ -75,7 +75,7 @@ class ModelCheckpoint(Callback):
             saved (``model.save_weights(filepath)``), else the full model
             is saved (``model.save(filepath)``).
         period: Interval (number of epochs) between checkpoints.
-
+        prefix: A string to put at the beginning of metric keys.
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply pytorch_lightning/callbacks/model_checkpoint.py:78]: Yes, should be removed. Noticed this yesterday. I was going to create an issue/PR for this.

File: tests/loggers/test_all.py [hunk 4]
```diff
@@ -316,3 +316,67 @@ def _test_logger_created_on_rank_zero_only(tmpdir, logger_class):
...
+            (1, ['tmp_epoch', 'tmp_test_acc', 'tmp_test_loss'])
+        ]
+        assert log_metric_names == expected
+    else:
+        expected = [
+            (0, ['tmp_epoch', 'tmp_train_some_val']),
+            (0, ['tmp_early_stop_on', 'tmp_epoch', 'tmp_val_acc']),
+            (1, ['tmp_epoch', 'tmp_test_acc', 'tmp_test_loss'])
+        ]
+        assert log_metric_names == expected
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply tests/loggers/test_all.py:382]: yep, can be done that way too. Just wanted to make sure that nothing else is updated or removed if any changes are made in the future. WDYT is checking that every string starts with that prefix is enough or should I keep it as it is?

File: tests/loggers/test_all.py (hunk 2 above)
Comment (rohitgr7) [reply tests/loggers/test_all.py]: @awaelchli can't get this working. Can you check, please??

**Comment by rohitgr7** [comment]
@awaelchli can't get the tests working. Not familiar with the error it's raising. Can you check, please?? Thanks.

**Comment by awaelchli** [comment]
ok, I try to find what's wrong :)

**Review by awaelchli — APPROVED** [review]
@rohitgr7 see my modifications to test. 
1. need to mock each 3rd party experiment object, because we don't install it in CI
2. each logger has slightly different interface
with the mocks it allows us to catch the exact call to the experiment object.

File: tests/loggers/test_all.py [hunk 5]
```diff
@@ -322,54 +330,43 @@ def test_logger_with_prefix_all(tmpdir, monkeypatch):
...
+        logger.log_metrics({"test": 1.0}, step=0)
+        logger.experiment.log_metrics.assert_called_once_with({"tmp-test": 1.0}, epoch=None, step=0)
 
+    # MLflow
     with mock.patch('pytorch_lightning.loggers.mlflow.mlflow'), \
          mock.patch('pytorch_lightning.loggers.mlflow.MlflowClient'):
-        _test_loggers_with_prefix(tmpdir, MLFlowLogger)
+        logger = _instantiate_logger(MLFlowLogger, save_idr=tmpdir, prefix=prefix)
+        logger.log_metrics({"test": 1.0}, step=0)
+        logger.experiment.log_metric.has_calls([call(ANY, "tmp-test", 1.0, ANY, step=0)])
```
Comment (rohitgr7) [inline tests/loggers/test_all.py:348]: great. Thanks, @awaelchli :smiley:. Need to study this mock tests stuff.

BTW why is this different from assertions for other loggers with `assert_called_once_with`?
Also, should it be `assert_has_calls`?

Comment (awaelchli) [reply tests/loggers/test_all.py]: yes, you are right. fixed it.
the `assert_called_once_with` accepts just a single call check, while `assert_has_calls` takes a list, which is useful checking the exact order of calls.
